pl_bolts/models/detection/yolo/yolo_module.py

```diff
@@ -0,0 +1,516 @@
+        Runs a forward pass through the network (all layers listed in `self.network`), and if
+        training targets are provided, computes the losses from the detection layers.
+
+        Detections are concatenated from the detection layers. Each image will produce
+        `N * num_anchors * grid_height * grid_width` detections, where `N` depends on the number of
+        detection layers. For one detection layer `N = 1`, and each detection layer increases it by
+        a number that depends on the size of the feature map on that layer. For example, if the
+        feature map is twice as wide and high as the grid, the layer will add four times more
+        features.
+
+        Args:
+            images: Images to be processed. Tensor of size
+                `[batch_size, num_channels, height, width]`.
+            targets: If set, computes losses from detection layers against these targets. A list of
+                dictionaries, one for each image.
+
+        Returns:
+            boxes (Tensor), confidences (Tensor), classprobs (Tensor), losses (Dict[str, Tensor]):
+                Detections, and if targets were provided, a dictionary of losses. The first
+                dimension of the detections is the index of the image in the batch and the second
+                dimension is the detection within the image. `boxes` contains the predicted
+                (x1, y1, x2, y2) coordinates, normalized to [0, 1].
+        """
+        outputs = []  # Outputs from all layers
+        detections = []  # Outputs from detection layers
+        losses = []  # Losses from detection layers
+
+        x = images
+        for module in self.network:
+            if isinstance(module, (RouteLayer, ShortcutLayer)):
+                x = module(x, outputs)
+            elif isinstance(module, DetectionLayer):
+                if targets is None:
+                    x = module(x)
+                    detections.append(x)
+                else:
+                    x, layer_losses = module(x, targets)
+                    detections.append(x)
+                    losses.append(layer_losses)
+            else:
+                x = module(x)
+
+            outputs.append(x)
+
+        def mean_loss(loss_name):
+            loss_tuple = tuple(layer_losses[loss_name] for layer_losses in losses)
+            return torch.stack(loss_tuple).sum() / images.shape[0]
+
+        detections = torch.cat(detections, 1)
+        boxes = detections[..., :4]
+        confidences = detections[..., 4]
+        classprobs = detections[..., 5:]
+
+        if targets is None:
+            return boxes, confidences, classprobs
+
+        losses = {loss_name: mean_loss(loss_name) for loss_name in losses[0].keys()}
+        return boxes, confidences, classprobs, losses
+
+    def configure_optimizers(self) -> Tuple[List, List]:
+        """Constructs the optimizer and learning rate scheduler."""
+        if self.optimizer == 'sgd':
+            optimizer = optim.SGD(
```

Comment:
What not pass `optimizer: Type[Optimizer] = torch.optim.Adam` instead of `str`? Wouldn't need the conditions here.

Reply:
I guess I wrote it this way because it was easy to add the constructor arguments to the CLI arguments using `argparse_utils.add_argparse_args()`, but you're right - it's a lot better to give the opportunity to provide any kind of optimizer object.